Support referring to distributed slices declared in sub-modules

As far as I can tell, if a distributed slice is declared anywhere beside the crate root, the element macro can't be used from any other scope.

Here's a simple example, adapted from the `distributed_slice` test:

```rust
mod declaration {
    use linkme::distributed_slice;

    #[distributed_slice]
    pub static SLICE: [i32] = [..];
}

mod usage {
    use linkme::distributed_slice;

    #[distributed_slice(crate::declaration::SLICE)]
    pub static N: i32 = 9;
}

mod alt_usage {
    use linkme::distributed_slice;

    use crate::declaration::SLICE;

    #[distributed_slice(SLICE)]
    pub static N: i32 = 9;
}
```

Both `usage` and `alt_usage` fail to compile:

```
$ cargo test
   Compiling linkmetest v0.1.0 (/home/daboross/linkmetest)
error[E0433]: failed to resolve: could not find `SLICE` in `declaration`
  --> src/main.rs:11:45
   |
11 |     #[distributed_slice(crate::declaration::SLICE)]
   |                                             ^^^^^ could not find `SLICE` in `declaration`

error: cannot find macro `SLICE!` in this scope
  --> src/main.rs:20:25
   |
20 |     #[distributed_slice(SLICE)]
   |                         ^^^^^
   |
   = help: have you added the `#[macro_use]` on the module/import?

warning: unused import: `crate::declaration::SLICE`
  --> src/main.rs:18:9
   |
18 |     use crate::declaration::SLICE;
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(unused_imports)]` on by default

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0433`.
error: Could not compile `linkmetest`.

To learn more, run the command again with --verbose.
```

I'm assuming this case simply isn't supported. But since this seems potentially quite useful, I'm opening this issue to track it.

---

I believe the cause for this is the usage of `macro_rules!` macros, and the fact that when exported using `#[macro_export]`, they're always exported from the crate root, regardless of the module they're declared in. This works as long as the static is also exported from the crate root, but as soon as it's declared in a sub-module, it's hard to make this work.

--- a/.travis.yml
+++ b/.travis.yml
@@ -7,8 +7,6 @@ rust:
 
 matrix:
   include:
-    - rust: 1.31.0
-      script: cargo test --tests
     - rust: 1.32.0
     - rust: nightly
       os: osx

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -9,6 +9,10 @@ documentation = "https://docs.rs/linkme"
 readme = "README.md"
 edition = "2018"
 
+[[test]]
+name = "module_2015"
+edition = "2015"
+
 [dependencies]
 linkme-impl = { version = "=0.1.5", path = "impl" }
 

--- a/README.md
+++ b/README.md
@@ -18,7 +18,7 @@ here.
 linkme = "0.1"
 ```
 
-*Supports rustc 1.31+*
+*Supports rustc 1.32+*
 
 <br>
 

--- a/impl/src/declaration.rs
+++ b/impl/src/declaration.rs
@@ -64,8 +64,8 @@ pub fn expand(input: TokenStream) -> TokenStream {
 
     let call_site = Span::call_site();
     let ident_str = ident.to_string();
-    let link_section_macro_dummy = format!("_linkme_macro_{}", ident);
-    let link_section_macro_dummy = Ident::new(&link_section_macro_dummy, call_site);
+    let link_section_macro_dummy_str = format!("_linkme_macro_{}", ident);
+    let link_section_macro_dummy = Ident::new(&link_section_macro_dummy_str, call_site);
 
     TokenStream::from(quote! {
         #(#attrs)*
@@ -102,9 +102,14 @@ pub fn expand(input: TokenStream) -> TokenStream {
             }
         };
 
-        #[derive(linkme::link_section_macro)]
+        #[macro_use]
+        mod #link_section_macro_dummy {
+            #[derive(linkme::link_section_macro)]
+            #[linkme_ident = #ident_str]
+            #[linkme_macro = #link_section_macro_dummy_str]
+            struct __linkme_dummy;
+        }
         #[doc(hidden)]
-        #[linkme_ident = #ident_str]
-        struct #link_section_macro_dummy;
+        #vis use #link_section_macro_dummy as #ident;
     })
 }

--- a/impl/src/declaration.rs
+++ b/impl/src/declaration.rs
@@ -66,6 +66,8 @@ pub fn expand(input: TokenStream) -> TokenStream {
     let ident_str = ident.to_string();
     let link_section_macro_dummy_str = format!("_linkme_macro_{}", ident);
     let link_section_macro_dummy = Ident::new(&link_section_macro_dummy_str, call_site);
+    let link_section_mod_dummy_str = format!("_linkme_module_{}", ident);
+    let link_section_mod_dummy = Ident::new(&link_section_mod_dummy_str, call_site);
 
     TokenStream::from(quote! {
         #(#attrs)*
@@ -103,12 +105,13 @@ pub fn expand(input: TokenStream) -> TokenStream {
         };
 
         #[macro_use]
-        mod #link_section_macro_dummy {
+        mod #link_section_mod_dummy {
             #[derive(linkme::link_section_macro)]
             #[linkme_ident = #ident_str]
             #[linkme_macro = #link_section_macro_dummy_str]
-            struct __linkme_dummy;
+            struct _linkme_dummy;
         }
+
         #[doc(hidden)]
         #vis use #link_section_macro_dummy as #ident;
     })

--- a/impl/src/derive.rs
+++ b/impl/src/derive.rs
@@ -7,13 +7,17 @@ use crate::linker;
 
 pub fn expand(input: DeriveInput) -> TokenStream {
     let mut linkme_ident = None;
+    let mut linkme_macro = None;
     for attr in input.attrs {
         if attr.path.is_ident("linkme_ident") {
             linkme_ident = parse_linkme_ident.parse2(attr.tokens).ok();
+        } else if attr.path.is_ident("linkme_macro") {
+            linkme_macro = parse_linkme_ident.parse2(attr.tokens).ok();
         }
     }
 
     let ident = linkme_ident.expect("attribute linkme_ident");
+    let ident_macro = linkme_macro.expect("attribute linkme_macro");
     let linux_section = linker::linux::section(&ident);
     let macos_section = linker::macos::section(&ident);
     let windows_section = linker::windows::section(&ident);
@@ -21,7 +25,7 @@ pub fn expand(input: DeriveInput) -> TokenStream {
     TokenStream::from(quote! {
         #[doc(hidden)]
         #[macro_export]
-        macro_rules! #ident {
+        macro_rules! #ident_macro {
             ($item:item) => {
                 #[used]
                 #[cfg_attr(any(target_os = "none", target_os = "linux"), link_section = #linux_section)]

--- a/impl/src/lib.rs
+++ b/impl/src/lib.rs
@@ -24,7 +24,7 @@ pub fn distributed_slice(args: TokenStream, input: TokenStream) -> TokenStream {
 }
 
 #[doc(hidden)]
-#[proc_macro_derive(link_section_macro, attributes(linkme_ident))]
+#[proc_macro_derive(link_section_macro, attributes(linkme_ident, linkme_macro))]
 pub fn link_section_macro(input: TokenStream) -> TokenStream {
     let expanded = derive::expand(parse_macro_input!(input));
     TokenStream::from(expanded)

--- a/tests/module.rs
+++ b/tests/module.rs
@@ -1,0 +1,18 @@
+mod declaration {
+    use linkme::distributed_slice;
+
+    #[distributed_slice]
+    pub static SLICE: [i32] = [..];
+
+    #[test]
+    fn test_mod_slice() {
+        assert!(!SLICE.is_empty());
+    }
+}
+
+mod usage {
+    use linkme::distributed_slice;
+
+    #[distributed_slice(super::declaration::SLICE)]
+    pub static N: i32 = 9;
+}

--- a/tests/module_2015.rs
+++ b/tests/module_2015.rs
@@ -1,0 +1,3 @@
+extern crate linkme;
+
+include!("module.rs");

